File: java/PracticaPDOO/src/main/java/deepspace/Dice.java
/**
 *
 * @author Alejandro Villanueva Prados
 */
package deepspace;

public class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    private java.util.Random generator;
    
    Dice() {
        NHANGARSPROB = 0.25f;
        NSHIELDSPROB = 0.25f;
        NWEAPONSPROB = 0.33f;
        FIRSTSHOTPROB = 0.5f;
        
        generator = new java.util.Random();
    }
    
    int initWithNHangars() {
        if(generator.nextFloat() > NHANGARSPROB)
            return 1;
        else 
            return 0;
    }
    
    int initWithNWeapons() {
        float prob = generator.nextFloat();
        if( prob < NWEAPONSPROB)
            return 1;
        else if( prob < 2*NWEAPONSPROB)
            return 2;
        else 
            return 3;
    }
    
    int initWithNShields() {
        if(generator.nextFloat() > NHANGARSPROB)
            return 1;
        else 
            return 0;
    }
    
    int whoStarts(int nPlayers) {
        return generator.nextInt(nPlayers);
    }
    
    GameCharacter firstShot() {
        if(generator.nextFloat() > FIRSTSHOTPROB)
            return GameCharacter.ENEMYSTARSHIP;
        else 
            return GameCharacter.SPACESTATION;
    }
    
    Boolean spaceStationMoves(float speed) {
        return generator.nextFloat() <= speed;
    }
}

File: ruby/Dice.rb
require_relative "GameCharacter"

module Deepspace
  class Dice
    def initialize
      @NHANGARSPROB = 0.25
      @NSHIELDSPROB = 0.25
      @NWEAPONSPROB = 0.33
      @FIRSTSHOTPROB = 0.5
      @EXTRAEFFICIENCYPROB = 0.8

      @generator = Random.new
    end

    def initWithNHangars
      if @generator.rand < @NHANGARSPROB
        0
      else
        1
      end
    end

    def initWithNWeapons
      rng = @generator.rand
      if rng < @NWEAPONSPROB
        1
      elsif rng < 2*@NWEAPONSPROB
        2
      else
        3
      end
    end

    def initWithNShields
      if @generator.rand < @NSHIELDSPROB
        0
      else
        1
      end
    end

    def whoStarts(nPlayers)
      @generator.rand(nPlayers)
    end

    def firstShot
      if @generator.rand < @FIRSTSHOTPROB
        GameCharacter::SPACESTATION
      else
        GameCharacter::ENEMYSTARSHIP
      end
    end

    def spaceStationMoves(speed)
      if @generator.rand < speed
        true
      else
        false
      end
    end

    def extraEfficiency
      @generator.rand < @EXTRAEFFICIENCYPROB
    end 
  end # class
end # module
